pulpcore/plugin/stages/content_stages.py

```diff
@@ -187,3 +187,55 @@ async def run(self):
         async for d_content in self.items():
             d_content.resolve()
             await self.put(d_content)
+
+
+class ContentAssociation(Stage):
+    """
+    A Stages API stage that associates content units with `new_version`.
+
+    This stage stores all content unit primary keys in memory before running. This is done to
+    compute the units already associated but not received from `self._in_q`. These units are passed
+    via `self._out_q` to the next stage as a :class:`django.db.models.query.QuerySet`.
+
+    This stage creates a ProgressReport named 'Associating Content' that counts the number of units
+    associated. Since it's a stream the total count isn't known until it's finished.
+
+    Args:
+        new_version (:class:`~pulpcore.plugin.models.RepositoryVersion`): The repo version this
+            stage associates content with.
+        args: unused positional arguments passed along to :class:`~pulpcore.plugin.stages.Stage`.
+        kwargs: unused keyword arguments passed along to :class:`~pulpcore.plugin.stages.Stage`.
+    """
+
+    def __init__(self, new_version, allow_delete, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.new_version = new_version
+        self.allow_delete = allow_delete
+
+    async def run(self):
+        """
+        The coroutine for this stage.
+
+        Returns:
+            The coroutine for this stage.
+        """
+        with ProgressReport(message="Associating Content", code="associating.content") as pb:
+            to_delete = set(self.new_version.content.values_list("pk", flat=True))
+            async for batch in self.batches():
+                to_add = set()
+                for d_content in batch:
+                    try:
+                        to_delete.remove(d_content.content.pk)
+                    except KeyError:
+                        to_add.add(d_content.content.pk)
+
+                if to_add:
+                    self.new_version.add_content(Content.objects.filter(pk__in=to_add))
+                    pb.increase_by(len(to_add))
+
+            if self.allow_delete and to_delete:
+                with ProgressReport(
+                    message="Un-Associating Content", code="unassociating.content"
+                ) as pb:
```

Comment:
I think, i would expect the progress report to exist if in mirror mode even, if there is nothing to delete.

Reply:
This is consistent with what it was doing before - the ContentUnassociation stage never ran at all unless mirror mode was active.

Reply:
What i mean would be something like:
```suggestion
            if self.allow_delete:
                with ProgressReport(
                    message="Un-Associating Content", code="unassociating.content"
                ) as pb:
                    if to_delete:
```

Reply:
I see what you mean.  Done.